src/main.rs

```diff
@@ -20,81 +24,103 @@ impl LinkedList {
         Self {
             head: None,
             last: None,
-            current_iter_elem: None,
         }
     }
 
-    fn add_before_head(&mut self, value: i32) {
-        if self.head.is_none() {
+    fn push_front(&mut self, value: i32) {
+        if let Some(head) = &self.head {
             self.head = Some(Rc::new(RefCell::new(Node {
                 data: value,
-                next: None,
+                next: Some(Rc::clone(head)),
             })));
         } else {
-            let copy_of_head = Rc::clone(self.head.as_ref().unwrap());
             self.head = Some(Rc::new(RefCell::new(Node {
                 data: value,
-                next: Some(Rc::clone(&copy_of_head)),
+                next: None,
             })));
         }
-        self.current_iter_elem = self.head.clone();
     }
 
-    fn add_after_n(&mut self, value: i32, mut n: i32) {
-        let mut current = self.head.clone();
+    fn add_after_n(&mut self, value: i32, index: i32) -> Result<(), String> {
+        if self.head.is_none() || self.size() < index {
+            return Err(format!("no item with index {}", index));
+        }
 
-        while n == 0 {
-            current = current.unwrap().borrow().next.clone();
-            n -= 1;
+        if self.size() == index {
+            self.push_back(value);
+            return Ok(());
         }
 
-        let tmp = current.clone().unwrap();
-        let mut ref_mut = tmp.borrow_mut();
-        let tmp_next = ref_mut.next.clone();
+        let mut current = self.head.clone();
+        for _ in 0..index {
+            if let Some(node) = current {
+                current = node.borrow().next.clone();
+            }
+        }
 
-        ref_mut.next = Some(Rc::new(RefCell::new(Node {
-            data: value,
-            next: tmp_next,
-        })))
-    }
+        if current.is_some() {
+            let tmp = current.clone().unwrap();
+            let mut ref_mut = tmp.borrow_mut();
+            let tmp_next = ref_mut.next.clone();
 
-    fn add_after_tail(&mut self, value: i32) {
-        if self.head.is_none() {
-            self.head = Some(Rc::new(RefCell::new(Node {
+            ref_mut.next = Some(Rc::new(RefCell::new(Node {
                 data: value,
-                next: None,
+                next: tmp_next,
             })));
-            self.current_iter_elem = self.head.clone();
-            self.last = self.head.clone();
+            Ok(())
         } else {
-            let mut current = self.head.clone();
-            while let Some(node) = current {
-                current = node.borrow().next.clone();
-                self.last = Some(node.clone());
+            Err(format!("can't insert after  item with index {}", index))
```

Comment:
```suggestion
            Err(format!("can't insert after  item with index {index}"))
```

`format!` раньше не поддерживал этот синтаксис, но давно уже поддерживает, так что советую использовать его